mllib/src/main/scala/org/apache/spark/ml/image/HadoopUtils.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.image
+
+import scala.language.existentials
+import scala.util.Random
+
+import org.apache.commons.io.FilenameUtils
+import org.apache.hadoop.conf.{Configuration, Configured}
+import org.apache.hadoop.fs.{Path, PathFilter}
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
+
+import org.apache.spark.sql.SparkSession
+
+private object RecursiveFlag {
+  /**
+   * Sets the spark recursive flag and then restores it.
+   *
+   * @param value Value to set
+   * @param spark Existing spark session
+   * @param f The function to evaluate after setting the flag
+   * @return Returns the evaluation result T of the function
+   */
+  def withRecursiveFlag[T](value: Boolean, spark: SparkSession)(f: => T): T = {
+    val flagName = FileInputFormat.INPUT_DIR_RECURSIVE
+    val hadoopConf = spark.sparkContext.hadoopConfiguration
+    val old = Option(hadoopConf.get(flagName))
+    hadoopConf.set(flagName, value.toString)
+    try f finally {
+      old match {
+        case Some(v) => hadoopConf.set(flagName, v)
+        case None => hadoopConf.unset(flagName)
+      }
+    }
+  }
+}
+
+/**
+ * Filter that allows loading a fraction of HDFS files.
+ */
+private class SamplePathFilter extends Configured with PathFilter {
```

Comment:
yes, I'm not sure about whether it will be deterministic even if we set a seed, but I can try to do that for now.  As @thunterdb suggested, we could use some sort of a hash on the filename - but I'm not sure on how I would make that implementation work with a specified ratio - could you give me more info on the design:

"I would prefer that we do not use a seed and that the result is deterministic, based for example on some hash of the file name, to make it more robust to future code changes. That being said, there is no fundamental issues with the current implementation and other developers may have differing opinions, so the current implementation is fine as far as I am concerned."

Reply:
Oh, it would be pretty simple to work it out this way (pseudocode). Note that a hash is a random variable between -2^31 and 2^31-1, so:
```
val fraction: Double = ??? // Something between 0 and 1
val pathname: String = ???
val hash = pathname.hashcode() // Could use some other other, more robust methods that return longs. 
val shouldKeep: Boolean = math.abs(hash) < (math.pow(2, 31) * fraction)
```

Reply:
Actually, I take this comment back, since it would break on some pathological cases such as all the names being the same. When users want some samples, they most probably want a result that is a fraction of the original, whatever it may contain.

@jkbradley do you prefer a something that may not be deterministic (using random numbers) or deterministic but not respecting the sampling ratio in pathological cases? The only way to do both that I can think of is deduplicating, which requires a shuffle.

Reply:
I would prefer determinism since that's a pretty important standard in Spark.  I could imagine either (a) using a file hash with a *global* random number or (b) using random numbers if we are certain about how PathFilters work.

For (a):
* Why is there a worry about duplicate filenames?  Is the full path not available?
* If you do hash filenames, then I wouldn't generate a random number for each row.  (If you're generating a random number per row, then why not just use that for sampling and skip the hash?)  You could generate a single random number on the driver and use that in the comparison with each hash.

For (b):
* If we knew how PathFilters were consumed, then we could presumably figure out a way to make this deterministic just by setting a random seed here. E.g., if a new PathFilter instance were instantiated to read each partition, then that would work.  But if PathFilters are shared across reads of multiple partitions, then partition ordering could cause problems with determinism.

Reply:
I don't believe readImages would ever return duplicate file names - even when reading from zip files the full path is included from the root.  I'm not sure what other pathological cases are possible - the hash function needs to be very good though - it needs to generate uniformly distributed random values (eg, if all values are above math.pow(2, 31) * 0.1234, then this won't work very well, or even if hashed values tend to be distributed more in some ranges than others) - I need to look into how the hash code is generated more

Reply:
It looks like the default hashCode is computed as:
s[0]*31^(n-1) + s[1]*31^(n-2) + ... + s[n-1]
where s[i] is the ith character of the string

Reply:
Ok, if we think that duplicate file names are not an issue, then I would prefer having using the deterministic hashing-based scheme. I am also happy to make this into another PR since this is a pretty small matter.

@imatiach-msft a good hashing function is certainly a concern, and we will need to make a trade off between performance and correctness. If we really want to make sure that it works as expected, the best is probably to use a cryptographic hash, like `SHA-256` (which has strong guarantees of the distribution of output values):
https://stackoverflow.com/questions/5531455/how-to-hash-some-string-with-sha256-in-java
We have `murmur3` in the Spark source code, but it is not cryptographic and does not come with as strong guarantees. For the sake of performance, we may want to use it eventually.

Once you have the digest, which is a byte array, then it can be converted to a long first (by taking 8 bytes from the whole digest):
https://stackoverflow.com/questions/4485128/how-do-i-convert-long-to-byte-and-back-in-java
and then you can convert this long to a double and compare it to the requested fraction:

```java
  Long l = new Long(...);
    double d = l.doubleValue();
   bool shouldKeep = d / (2<<31) <= requestedFraction
```